inception-workload-dynamic/src/main/java/de/tudarmstadt/ukp/inception/workload/dynamic/annotation/DynamicAnnotatorWorkflowActionBarItemGroup.java

```diff
@@ -89,48 +101,91 @@ public String getStateClass()
             User user = state.getUser();
             Project project = state.getProject();
             SourceDocument document = state.getDocument();
 
-            AnnotationDocument annotationDocument = documentService
-                .getAnnotationDocument(document, user);
+            annotationDocumentList = documentService.listAnnotationDocuments(project, user);
+
+            AnnotationDocument annotationDocument = documentService.getAnnotationDocument(document,
+                    user);
 
             documentService.transitionAnnotationDocumentState(annotationDocument,
-                ANNOTATION_IN_PROGRESS_TO_ANNOTATION_FINISHED);
-
-            //Go through all documents in a random order and check if there is a Annotation document
-            //with the state NEW
-            String query =  "FROM SourceDocument " +
-                            "WHERE project = :project " +
-                            "ORDER BY rand()";
-            for (SourceDocument doc: entityManager.createQuery(query,SourceDocument.class)
-                .setParameter("project", project).getResultList()) {
-                //Check if it even exists or is state NEW
-                if (documentService.listAnnotatableDocuments(project,user).get(doc) == null ||
-                    documentService.listAnnotatableDocuments(project,user).get(doc).
-                        getState().equals(NEW)) {
-                    getAnnotationPage().getModelObject().setDocument(doc, documentService.
-                        listSourceDocuments(project));
-                    //This document had the state NEW, load it
-                    getAnnotationPage().actionLoadDocument(_target);
-                    _target.add(page);
+                    ANNOTATION_IN_PROGRESS_TO_ANNOTATION_FINISHED);
+
+            for (AnnotationDocument anno : annotationDocumentList) {
+                // There was one in progress, load it
+                if (anno.getState().equals(IN_PROGRESS)) {
+                    getAnnotationPage().getModelObject().setDocument(anno.getDocument(),
+                            documentService.listSourceDocuments(project));
+                    Optional<AjaxRequestTarget> target = RequestCycle.get()
+                            .find(AjaxRequestTarget.class);
+                    getAnnotationPage().actionLoadDocument(target.orElse(null));
                     return;
                 }
             }
 
-            //Check if no new document has been selected, return to homepage
-            if (getAnnotationPage().getModelObject().getDocument().equals(document)) {
-                getAnnotationPage().setResponsePage(getAnnotationPage().getApplication().
-                    getHomePage());
-                getSession().info("There are no more documents to annotate available for you. Please contact your project supervisor.");
+            if ((workloadManagementService.getTraits(project).split(",")[0]).equals("randomized")) {
```

Comment:
Traits are by convention JSON objects and we use a common pattern for accessing them consisting of the methods `readTraits` and `writeTraits` (cf. e.g. the LayerSupport_ImplBase or FeatureSupport and its implementations). And we do not create the JSON manually, but rather from a `XXXTraits` (c.f. StringFeatureTraits) class using Jackson. I believe your case, these methods should go into the DynamicWorkloadExtension and MatrixWorkloadExtension classes. The readTraits method take a database entity with a traits field, read the field, parse it into an extension-specific data object, and returns it. The writeTraits does it the other way around. The database entity should be obtained from the DAO class - which in your case would be WorkloadManagementService.getOrCreateWorkloadManagerConfiguration() and after a change, it should be written using the corresponding write method in the DAO, so in your case WorkloadManagementService.setWorkloadManagerConfiguration()